File: server/sonar-ws-client/src/main/java/org/sonar/wsclient/component/Component.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.wsclient.component;

import org.sonar.wsclient.unmarshallers.JsonUtils;

import javax.annotation.CheckForNull;

import java.util.Map;

public class Component {
  private final Map json;

  public Component(Map json) {
    this.json = json;
  }

  public String qualifier() {
    return JsonUtils.getString(json, "qualifier");
  }

  public String key() {
    return JsonUtils.getString(json, "key");
  }

  /**
   * @since 4.2
   */
  public Long id() {
    return JsonUtils.getLong(json, "id");
  }

  public String name() {
    return JsonUtils.getString(json, "name");
  }

  @CheckForNull
  public String longName() {
    return JsonUtils.getString(json, "longName");
  }

  /**
   * @since 4.2
   */
  @CheckForNull
  public Long subProjectId() {
    return JsonUtils.getLong(json, "subProjectId");
  }

  /**
   * @since 4.2
   */
  @CheckForNull
  public Long projectId() {
    return JsonUtils.getLong(json, "projectId");
  }

  /**
   * @since 4.2
   */
  @CheckForNull
  public String path() {
    return JsonUtils.getString(json, "path");
  }

}

File: server/sonar-web/src/main/js/helpers/mocks/component.ts
/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier } from '~sonar-aligned/types/component';
import { MetricKey } from '~sonar-aligned/types/metrics';
import { TreeComponent, Visibility } from '../../types/component';
import { Component, ComponentMeasure, ComponentMeasureEnhanced } from '../../types/types';
import { mockMeasureEnhanced } from '../testMocks';

export function mockComponent(overrides: Partial<Component> = {}): Component {
  return {
    breadcrumbs: [],
    key: 'my-project',
    name: 'MyProject',
    qualifier: ComponentQualifier.Project,
    qualityGate: { isDefault: true, key: '30', name: 'Sonar way' },
    qualityProfiles: [
      {
        deleted: false,
        key: 'my-qp',
        language: 'ts',
        name: 'Sonar way',
      },
    ],
    tags: [],
    ...overrides,
  };
}

export function mockTreeComponent(overrides: Partial<TreeComponent>): TreeComponent {
  return {
    key: 'my-key',
    qualifier: ComponentQualifier.Project,
    name: 'component',
    visibility: Visibility.Public,
    ...overrides,
  };
}

export function mockComponentMeasure(
  file = false,
  overrides: Partial<ComponentMeasure> = {},
): ComponentMeasure {
  if (file) {
    return {
      key: 'foo:src/index.tsx',
      name: 'index.tsx',
      qualifier: ComponentQualifier.File,
      path: 'src/index.tsx',
      measures: [{ metric: MetricKey.bugs, value: '1', bestValue: false }],
      ...overrides,
    };
  }
  return {
    key: 'foo',
    name: 'Foo',
    qualifier: ComponentQualifier.Project,
    measures: [{ metric: MetricKey.bugs, value: '12', bestValue: false }],
    ...overrides,
  };
}

export function mockComponentMeasureEnhanced(
  overrides: Partial<ComponentMeasureEnhanced> = {},
): ComponentMeasureEnhanced {
  return {
    ...mockComponentMeasure(false, overrides as ComponentMeasure),
    leak: undefined,
    measures: [mockMeasureEnhanced()],
    value: undefined,
    ...overrides,
  };
}
